lib/charcoal/jsonp.rb

```diff
@@ -38,8 +42,8 @@ def add_jsonp_callback
       yield
 
       if response.status.to_s.starts_with?('200') && jsonp_request?
-        response.body = "#{params[:callback]}(#{response.body})"
         response.content_type = "application/javascript"
+        response.body = "#{params[:callback]}(#{response.body.to_s})"
```

Comment:
The suggested solution doesn’t work for me. First, the cookies are committed:
```
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/middleware/cookies.rb:17:in `commit_cookie_jar!'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:416:in `before_sending'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:187:in `block in sending!'
/Users/bquorning/.rubies/ruby-2.2.3/lib/ruby/2.2.0/monitor.rb:211:in `mon_synchronize'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:186:in `sending!'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:114:in `each'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:100:in `body'
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/http/response.rb:282:in `body'
/Users/bquorning/.gem/ruby/2.2.3/bundler/gems/charcoal-9693731ec285/lib/charcoal/jsonp.rb:46:in `add_jsonp_callback'
...
/Users/bquorning/.gem/ruby/2.2.3/gems/rack-2.0.0.alpha/lib/rack/session/abstract/id.rb:220:in `context'
/Users/bquorning/.gem/ruby/2.2.3/gems/rack-2.0.0.alpha/lib/rack/session/abstract/id.rb:214:in `call'
```

And then, we try to set the cookies:
```
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/middleware/cookies.rb:465:in `[]='
/Users/bquorning/.gem/ruby/2.2.3/gems/actionpack-5.0.0.beta1/lib/action_dispatch/middleware/session/cookie_store.rb:117:in `set_cookie'
/Users/bquorning/.gem/ruby/2.2.3/gems/rack-2.0.0.alpha/lib/rack/session/abstract/id.rb:351:in `commit_session'
/Users/bquorning/.gem/ruby/2.2.3/gems/rack-2.0.0.alpha/lib/rack/session/abstract/id.rb:222:in `context'
/Users/bquorning/.gem/ruby/2.2.3/gems/rack-2.0.0.alpha/lib/rack/session/abstract/id.rb:214:in `call'
```

This would give me a 'RuntimeError: can't modify frozen Hash'.

I hacked together a way to read the stream buffer directly, without closing the response, but it’s ugly as heck:

```rb
body = response.stream.instance_variable_get(:@buf).join
response.body = "#{params[:callback]}(#{body})"
```

This also, obviously, doesn’t work with streaming. Perhaps, before `yield`, we could substitute the buffer for an object that we are allowed to read from?